src/codegen/templates/_library_interpreter.py.mako

```diff
@@ -156,6 +187,127 @@ class LibraryInterpreter(BaseInterpreter):
             return 'Failed to retrieve error description.'
         return error_buffer.value.decode(lib_importer.encoding)
 
+    ## internal_read_analog_waveform_ex has special handling for waveform attributes and callbacks
+    def internal_read_analog_waveform_ex(
+        self,
+        task_handle: object,
+        num_samps_per_chan: int,
+        timeout: float,
+        waveform: AnalogWaveform[numpy.float64]
+    ) -> None:
+        """Read an analog waveform with timing and attributes."""
+        assert isinstance(task_handle, TaskHandle)
+        samps_per_chan_read = ctypes.c_int()
+
+        cfunc = lib_importer.windll.DAQmxInternalReadAnalogWaveformEx
+        if cfunc.argtypes is None:
+            with cfunc.arglock:
+                if cfunc.argtypes is None:
+                    cfunc.argtypes = [
+                        TaskHandle,
+                        ctypes.c_int,
+                        ctypes.c_double,
+                        ctypes.c_int,
+                        wrapped_ndpointer(dtype=numpy.int64, flags=("C", "W")),
+                        wrapped_ndpointer(dtype=numpy.int64, flags=("C", "W")),
+                        ctypes.c_uint,
+                        CSetWfmAttrCallbackPtr,
+                        ctypes.c_void_p,
+                        wrapped_ndpointer(dtype=numpy.float64, flags=("C", "W")),
+                        ctypes.c_uint,
+                        ctypes.POINTER(ctypes.c_int),
+                        ctypes.POINTER(c_bool32),
+                    ]
+
+        t0_array = numpy.zeros(1, dtype=numpy.int64)
+        dt_array = numpy.zeros(1, dtype=numpy.int64)
+
+        def set_wfm_attr_callback(
+            channel_index: int,
+            attribute_name: str,
+            attribute_type: WfmAttrType,
+            value: ExtendedPropertyValue,
+            callback_data: object,
+        ) -> int:
+            waveform.extended_properties[attribute_name] = value
+            return 0
+
+        error_code = cfunc(
+            task_handle,
+            num_samps_per_chan,
+            timeout,
+            FillMode.GROUP_BY_CHANNEL.value,
+            t0_array,
+            dt_array,
+            t0_array.size,
+            self._get_wfm_attr_callback_ptr(set_wfm_attr_callback),
+            None,
+            waveform.raw_data,
+            waveform.raw_data.size,
+            ctypes.byref(samps_per_chan_read),
```

Comment:
Looking at MSDaqDotNET reminds me of another very important case; short/partial reads.

Reading N samples may return fewer than N samples. When this happens, the read function should resize the waveform to indicate how many samples were actually read, not how many samples it tried to read.

A short read may also be accompanied by an error. If this happens, the exception indicates the number of samples read. If the read function is returning a new waveform, the data is lost, but if the read function is reading into an existing waveform, the data is available in the waveform.